### PR TITLE
version: use OSS repo's version when building.

### LINKED ISSUE #841: Use open source repo as source of truth for versioning?
Currently we have a closed source meta-repo that submodules in a bunch of other git repos. We use its git version (and commits past last tag via `git describe`) to determine the version number of official releases.

We were discussing changing our releases to use the version in this open source repo instead.

That does mean if there there are other-repo-only changes (e.g. Windows client or Android client) that use this same version of tailscale/tailscale, they'd have the same version number. But those are rare and we can distinguish those by having an extra git suffix. Currently our full version number is of the form: `1.0.5-gXXXXXXX` where `XXXXXXX` is the git hash of our closed source meta repo. We could instead make that, say, `-gXXXXXXX_XXXXXXX` where the `_` (or some other character) separates out the git hashes of the two repos.

/cc @apenwarr @danderson @crawshaw @eliasnaur @alexbrainman

### PR DESCRIPTION
**Pull request by danderson**
When building with redo, also include the git commit hash
from the proprietary repo, so that we have a precise commit
that identifies all build info (including Go toolchain version).

Add a top-level build script demonstrating to downstream distros
how to burn the right information into builds.

Adjust `tailscale version` to print commit hashes when available.

Fixes #841.

---
This change is [Reviewable](https://reviewable.io/reviews/tailscale/tailscale/858)

### CODE REVIEW

**Comment by danderson** [comment]
Followups: print the right versions in windows, xcode and android.

Problem: with this logic, current head's build version is 1.1.290. The current versions on pkgs are 1.1.{540,556}. So, we need to either break current unstable, or insert some version hack specific to 1.1 (e.g. s/290/1290/).

**Comment by danderson** [comment]
Decided to go with 1000+patch for 1.1.x, so the next build after submitting this change will be 1.1.1291, and sequencing will be preserved. PTAL? Merging tomorrow.

File: build_dist.sh [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+# issues.
+#
+# If you're packaging Tailscale for a distro, please consider using
+# this script, or executing equivalent commands in your
+# distro-specific build system.
+
+set -euo pipefail
+
+describe=$(git describe --long --abbrev=9)
+commit=$(git describe --dirty --exclude "*" --always --abbrev=200)
```
Comment (bradfitz) [inline build_dist.sh:15]: why 200? I guess you just mean "something big so don't abbreviate it"?

File: version/xversion.go [hunk 2]
```diff
@@ -9,3 +9,5 @@ package version
 // Replaced at build time with the Go linker flag -X.
```
Comment (bradfitz) [inline version/xversion.go:9]: Reference build_dist.sh and version.go for docs on what these fields are?

File: version/version.go [hunk 3]
```diff
@@ -7,5 +7,29 @@
 // Package version provides the version that the binary was built at.
 package version
 
+// LONG is a full version number for this build, of the form
+// "x.y.z-commithash", or "date.yyyymmdd" if no actual version was
```
Comment (bradfitz) [inline version/version.go:11]: are we going to be appending a further `-gXXXXXX` to the `-commithash` for the corp repo's hash?

File: build_dist.sh (hunk 1 above)
Comment (danderson) [reply build_dist.sh:15]: "I can't remember what length a sha1 is, and also I don't want it to bork when hash agility finally shows up in git", yeah.

File: version/version.go (hunk 3 above)
Comment (danderson) [reply version/version.go:11]: As discussed, we're going for `1.2.3-tTAILSCALE_HASH-oOTHER_HASH`. The -o part is omitted in pure OSS builds.

**Pull request merged**